python/lib/core/dmod/core/data_domain_detectors.py

```diff
@@ -0,0 +1,468 @@
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from functools import reduce
+from pathlib import Path
+from typing import TypeVar, Union, Dict, Type, Set, List, Optional
+
+from .common.reader import ReadSeeker
+from .dataset import Dataset
+from .exception import DmodRuntimeError
+from .meta_data import DataDomain, DataFormat, StandardDatasetIndex, DiscreteRestriction
+
+DataItem = TypeVar('DataItem', bound=Union[bytes, ReadSeeker, Path])
+DataCollection = TypeVar('DataCollection', bound=Union[Dataset, Dict[str, Union[bytes, ReadSeeker]], Path])
+
+
+class AbstractDomainDetector(ABC):
+    """ Abstraction for something that will automatically detect a :class:`DataDomain` for some data. """
+
+    @abstractmethod
+    def detect(self, **kwargs) -> DataDomain:
+        """
+        Detect and return the data domain.
+
+        Parameters
+        ----------
+        kwargs
+            Optional kwargs applicable to the subtype, which may enhance or add to the domain detection and generation
+            capabilities, but which should not be required to produce a valid domain.
+
+        Returns
+        -------
+        DataDomain
+            The detected domain.
+
+        Raises
+        ------
+        DmodRuntimeError
+            If it was not possible to properly detect the domain.
+        """
+        pass
+
+
+class ItemDataDomainDetectorRegistry:
+    """ A singleton registry in which to track the subtypes of :class`ItemDataDomainDetector`. """
+
+    _instance = None
+
+    @classmethod
+    def get_instance(cls) -> ItemDataDomainDetectorRegistry:
+        """ Get the singleton registry instance. """
+        if cls._instance is None:
+            cls._instance = ItemDataDomainDetectorRegistry()
+        return cls._instance
+
+    def __init__(self):
+        if self._instance is not None:
+            raise RuntimeError(f"Attempting to create second {self.__class__.__name__} instance!")
+        self._detectors: Dict[str, Type[ItemDataDomainDetector]] = dict()
```

Comment:
No probably not. My thinking was to avoid constraints to external development of subtypes by providing a mechanism to avoid collisions on registration name if two classes of the same name were created. But that probably is overthinking more than actually usefulness.

Reply:
If you just use the `type` and not some string identifier, I think you avoid collisions by design, yeah? I think that would still support the external subtype use case.

Reply:
Yeah, this makes sense.  I've made this change, though it was part of a larger overhaul.  It made sense to change/remove some other things at the same time.